File: utils/pattern_finder.py
import numpy as np
from scipy.signal import convolve2d
import os
import time
import hashlib


class GameOfLife:
    def __init__(self, width, height, wrap=False):
        self.wrap = wrap
        self.grid = np.zeros((height, width), dtype=int)

    def step(self):
        # Compute the number of neighbors using convolution
        kernel = np.array([[1, 1, 1], [1, 0, 1], [1, 1, 1]])
        neighbors = convolve2d(
            self.grid, kernel, mode="same", boundary="wrap" if self.wrap else "fill"
        )

        # Apply the rules of Game of Life
        born = (self.grid == 0) & (neighbors == 3)
        survive = (self.grid == 1) & ((neighbors == 2) | (neighbors == 3))

        self.grid = np.where(born | survive, 1, 0)

    def dump(self):
        result = []
        for y in range(self.grid.shape[0]):
            result.append(
                "".join(
                    "#" if self.grid[y, x] else "." for x in range(self.grid.shape[1])
                )
            )
        return result


class PatternFinder:
    def __init__(self, width, height, attempts=1000):
        self.width = width
        self.height = height
        self.attempts = attempts

        # Ensure "results" directory exists
        if not os.path.exists("results"):
            os.makedirs("results")

    def random_initialize(self, game):
        game.grid = np.random.randint(2, size=(self.height, self.width))

    def find_unstable_patterns(self):
        max_length = 0
        last_print_time = time.time()

        for attempt in range(self.attempts):
            game = GameOfLife(self.width, self.height, wrap=True)
            self.random_initialize(game)

            seen_states = set()
            initial_pattern = game.dump()

            for gen in range(100000):
                current_state = tuple(map(tuple, game.grid))
                if current_state in seen_states:
                    break
                seen_states.add(current_state)
                game.step()
                length = len(seen_states)
                if length > max_length:
                    max_length = length
                if time.time() - last_print_time >= 1:  # Print every second
                    print(
                        f"Progress: {attempt}/{self.attempts}, Max length: {max_length}"
                    )
                    last_print_time = time.time()

            if len(seen_states) >= 500:
                self.save_pattern(initial_pattern, len(seen_states))

    def save_pattern(self, pattern, length):
        hash = hashlib.sha256("\n".join(pattern).encode("utf-8")).hexdigest()[0:8]
        filename = f"results/pattern_{length:09}_{hash}.txt"
        with open(filename, "w") as f:
            for line in pattern:
                f.write(line + "\n")


# Usage
finder = PatternFinder(8, 32, attempts=100000)
finder.find_unstable_patterns()

File: utils/pattern_finder.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <random>
#include <ctime>
#include <fstream>
#include <sstream>
#include <chrono>

class GameOfLife {
    std::vector<std::vector<int>> grid;
    bool wrap;
    int width;
    int height;

public:
    GameOfLife(int w, int h, bool wr = false) : width(w), height(h), wrap(wr) {
        grid.resize(height, std::vector<int>(width, 0));
    }

    void randomInitialize() {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dist(0, 1);

        for (int i = 0; i < height; i++)
            for (int j = 0; j < width; j++)
                grid[i][j] = dist(gen);
    }

    void step() {
        std::vector<std::vector<int>> newGrid(height, std::vector<int>(width, 0));

        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                int livingNeighbors = 0;
                for (int x = -1; x <= 1; x++) {
                    for (int y = -1; y <= 1; y++) {
                        if (x == 0 && y == 0) continue;

                        int ni = i + x;
                        int nj = j + y;

                        if (wrap) {
                            ni = (ni + height) % height;
                            nj = (nj + width) % width;
                        }

                        if (ni >= 0 && nj >= 0 && ni < height && nj < width) {
                            livingNeighbors += grid[ni][nj];
                        }
                    }
                }

                if (grid[i][j] == 1 && (livingNeighbors == 2 || livingNeighbors == 3))
                    newGrid[i][j] = 1;
                else if (grid[i][j] == 0 && livingNeighbors == 3)
                    newGrid[i][j] = 1;
            }
        }

        grid = newGrid;
    }

    std::vector<std::vector<int>> getGrid() const {
        return grid;
    }
};

int main() {
    const int attempts = 1000000;
    int width = 8;
    int height = 32;
    int maxIterations = 100000;

    std::unordered_set<std::string> seenStates;
    int maxFound = 0;

    auto lastPrint = std::chrono::system_clock::now();

    for (int attempt = 0; attempt < attempts; attempt++) {
        GameOfLife game(width, height, true);
        game.randomInitialize();

        std::vector<std::vector<int>> initialState = game.getGrid(); // Save initial state

        seenStates.clear();

        int iteration;
        for (iteration = 0; iteration < maxIterations; iteration++) {
            std::stringstream ss;
            for (const auto& row : game.getGrid()) {
                for (int cell : row) {
                    ss << cell;
                }
                ss << "\n";
            }
            std::string currentState = ss.str();

            if (seenStates.find(currentState) != seenStates.end()) {
                break;
            }

            seenStates.insert(currentState);
            game.step();

            // Check if it's time to print progress
            auto now = std::chrono::system_clock::now();
            std::chrono::duration<double> elapsedSeconds = now - lastPrint;

            if (elapsedSeconds.count() >= 1.0) { // If more than 1 second has passed
                std::cout << "Progress: " << attempt + 1 << "/" << attempts;
                std::cout << " | Max length found: " << maxFound << std::endl;
                lastPrint = now;
            }
        }

        if (iteration > maxFound) {
            maxFound = iteration;
        }

        if (iteration >= 600) {
            std::ofstream outfile("results/pattern_" + std::to_string(iteration) + ".txt");
            for (const auto& row : initialState) { // Use the saved initial state
                for (int cell : row) {
                    outfile << (cell ? "#" : "."); // Use # for life and . for empty
                }
                outfile << "\n";
            }
            outfile.close();
        }
    }

    return 0;
}
